src/execution_plan/ops/op_node_hash_join.c

```diff
@@ -0,0 +1,244 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "op_node_hash_join.h"
+#include "../../util/arr.h"
+#include "../../util/qsort.h"
+#include <assert.h>
+
+#define NO_INTERSECTION -1
+
+/* Determins order between two records by inspecting
+ * node stored at postion intersect_node_idx. */
+static bool _record_islt(Record l, Record r, uint intersect_node_idx) {
+    Node *ln = Record_GetNode(l, intersect_node_idx);
+    Node *rn = Record_GetNode(r, intersect_node_idx);
+    return ENTITY_GET_ID(ln) < ENTITY_GET_ID(rn);
+}
+
+/* `intersect_node_idx` is an actual variable in the caller function.
+ * Using it in a macro like this is rather ugly,
+ * but the macro passed to QSORT must accept only 2 arguments. */
+#define RECORD_SORT_ON_ENTRY(a, b) (_record_islt((*a), (*b), intersect_node_idx))
+
+/* Retrive the next intersecting record
+ * if such exists, otherwise returns NULL. */
+static Record _get_intersecting_record(OpNodeHashJoin *op) {
+    // No more intersecting records.
+    if(op->intersect_idx == NO_INTERSECTION ||
+       op->number_of_intersections == 0) return NULL;
+
+    Record cr = op->cached_records[op->intersect_idx];
+    
+    // Update intersection trackers.
+    op->intersect_idx++;
+    op->number_of_intersections--;
+
+    return cr;
+}
+
+static uint64_t _binarySearch(const OpNodeHashJoin *op, NodeID id) {
```

Comment:
This is the find-rightmost variant of the binary search algorithm; the name should probably reflect that.